## Final task

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, OrdinalEncoder
import matplotlib.pyplot as plt

In [2]:
data = pd.read_csv('responses.csv')

In [3]:
data.shape

(1010, 150)

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1010 entries, 0 to 1009
Columns: 150 entries, Music to House - block of flats
dtypes: float64(134), int64(5), object(11)
memory usage: 1.2+ MB


In [5]:
data.describe()

Music  Slow songs or fast songs        Dance         Folk  \
count  1007.000000               1008.000000  1006.000000  1005.000000   
mean      4.731877                  3.328373     3.113320     2.288557   
std       0.664049                  0.833931     1.170568     1.138916   
min       1.000000                  1.000000     1.000000     1.000000   
25%       5.000000                  3.000000     2.000000     1.000000   
50%       5.000000                  3.000000     3.000000     2.000000   
75%       5.000000                  4.000000     4.000000     3.000000   
max       5.000000                  5.000000     5.000000     5.000000   

           Country  Classical music      Musical          Pop         Rock  \
count  1005.000000      1003.000000  1008.000000  1007.000000  1004.000000   
mean      2.123383         2.956132     2.761905     3.471698     3.761952   
std       1.076136         1.252570     1.260845     1.161400     1.184861   
min       1.000000         1.000000     1.000000     1.000000     1.000000   
25%       1.000000         2.000000     2.000000     3.000000     3.000000   
50%       2.000000         3.000000     3.000000     4.000000     4.000000   
75%       3.000000         4.000000     4.000000     4.000000     5.000000   
max       5.000000         5.000000     5.000000     5.000000     5.000000   

       Metal or Hardrock  ...  Shopping centres  Branded clothing  \
count        1007.000000  ...       1008.000000       1008.000000   
mean            2.361470  ...          3.234127          3.050595   
std             1.372995  ...          1.323062          1.306321   
min             1.000000  ...          1.000000          1.000000   
25%             1.000000  ...          2.000000          2.000000   
50%             2.000000  ...          3.000000          3.000000   
75%             3.000000  ...          4.000000          4.000000   
max             5.000000  ...          5.000000          5.000000   

       Entertainment spending  Spending on looks  Spending on gadgets  \
count             1007.000000        1007.000000          1010.000000   
mean                 3.201589           3.106256             2.870297   
std                  1.188947           1.205368             1.284970   
min                  1.000000           1.000000             1.000000   
25%                  2.000000           2.000000             2.000000   
50%                  3.000000           3.000000             3.000000   
75%                  4.000000           4.000000             4.000000   
max                  5.000000           5.000000             5.000000   

       Spending on healthy eating          Age      Height      Weight  \
count                  1008.00000  1003.000000  990.000000  990.000000   
mean                      3.55754    20.433699  173.514141   66.405051   
std                       1.09375     2.828840   10.024505   13.839561   
min                       1.00000    15.000000   62.000000   41.000000   
25%                       3.00000    19.000000  167.000000   55.000000   
50%                       4.00000    20.000000  173.000000   64.000000   
75%                       4.00000    22.000000  180.000000   75.000000   
max                       5.00000    30.000000  203.000000  165.000000   

       Number of siblings  
count         1004.000000  
mean             1.297809  
std              1.013348  
min              0.000000  
25%              1.000000  
50%              1.000000  
75%              2.000000  
max             10.000000  

[8 rows x 139 columns]

In [6]:
data.head()

Music  Slow songs or fast songs  Dance  Folk  Country  Classical music  \
0    5.0                       3.0    2.0   1.0      2.0              2.0   
1    4.0                       4.0    2.0   1.0      1.0              1.0   
2    5.0                       5.0    2.0   2.0      3.0              4.0   
3    5.0                       3.0    2.0   1.0      1.0              1.0   
4    5.0                       3.0    4.0   3.0      2.0              4.0   

   Musical  Pop  Rock  Metal or Hardrock  ...   Age  Height  Weight  \
0      1.0  5.0   5.0                1.0  ...  20.0   163.0    48.0   
1      2.0  3.0   5.0                4.0  ...  19.0   163.0    58.0   
2      5.0  3.0   5.0                3.0  ...  20.0   176.0    67.0   
3      1.0  2.0   2.0                1.0  ...  22.0   172.0    59.0   
4      3.0  5.0   3.0                1.0  ...  20.0   170.0    59.0   

   Number of siblings  Gender  Left - right handed                Education  \
0                 1.0  female         right handed  college/bachelor degree   
1                 2.0  female         right handed  college/bachelor degree   
2                 2.0  female         right handed         secondary school   
3                 1.0  female         right handed  college/bachelor degree   
4                 1.0  female         right handed         secondary school   

   Only child  Village - town  House - block of flats  
0          no         village          block of flats  
1          no            city          block of flats  
2          no            city          block of flats  
3         yes            city          house/bungalow  
4          no         village          house/bungalow  

[5 rows x 150 columns]

In [7]:
# data = data.dropna(inplace = True)
for i in data.columns:    
    data[i] = data[i].fillna(data[i].mode()[0])

In [8]:
data.describe()

Music  Slow songs or fast songs        Dance         Folk  \
count  1010.000000               1010.000000  1010.000000  1010.000000   
mean      4.732673                  3.327723     3.112871     2.287129   
std       0.663222                  0.833232     1.168268     1.136271   
min       1.000000                  1.000000     1.000000     1.000000   
25%       5.000000                  3.000000     2.000000     1.000000   
50%       5.000000                  3.000000     3.000000     2.000000   
75%       5.000000                  4.000000     4.000000     3.000000   
max       5.000000                  5.000000     5.000000     5.000000   

           Country  Classical music      Musical          Pop         Rock  \
count  1010.000000      1010.000000  1010.000000  1010.000000  1010.000000   
mean      2.122772         2.956436     2.762376     3.473267     3.769307   
std       1.073501         1.248223     1.259639     1.160029     1.185161   
min       1.000000         1.000000     1.000000     1.000000     1.000000   
25%       1.000000         2.000000     2.000000     3.000000     3.000000   
50%       2.000000         3.000000     3.000000     4.000000     4.000000   
75%       3.000000         4.000000     4.000000     4.000000     5.000000   
max       5.000000         5.000000     5.000000     5.000000     5.000000   

       Metal or Hardrock  ...  Shopping centres  Branded clothing  \
count        1010.000000  ...       1010.000000       1010.000000   
mean            2.357426  ...          3.233663          3.050495   
std             1.372955  ...          1.321791          1.305028   
min             1.000000  ...          1.000000          1.000000   
25%             1.000000  ...          2.000000          2.000000   
50%             2.000000  ...          3.000000          3.000000   
75%             3.000000  ...          4.000000          4.000000   
max             5.000000  ...          5.000000          5.000000   

       Entertainment spending  Spending on looks  Spending on gadgets  \
count             1010.000000        1010.000000          1010.000000   
mean                 3.200990           3.105941             2.870297   
std                  1.187229           1.203589             1.284970   
min                  1.000000           1.000000             1.000000   
25%                  2.000000           2.000000             2.000000   
50%                  3.000000           3.000000             3.000000   
75%                  4.000000           4.000000             4.000000   
max                  5.000000           5.000000             5.000000   

       Spending on healthy eating          Age       Height       Weight  \
count                 1010.000000  1010.000000  1010.000000  1010.000000   
mean                     3.558416    20.423762   173.444554    66.278218   
std                      1.092842     2.821521     9.936737    13.730769   
min                      1.000000    15.000000    62.000000    41.000000   
25%                      3.000000    19.000000   167.000000    56.000000   
50%                      4.000000    20.000000   172.000000    64.000000   
75%                      4.000000    22.000000   180.000000    75.000000   
max                      5.000000    30.000000   203.000000   165.000000   

       Number of siblings  
count          1010.00000  
mean              1.29604  
std               1.01059  
min               0.00000  
25%               1.00000  
50%               1.00000  
75%               2.00000  
max              10.00000  

[8 rows x 139 columns]

In [9]:
data.shape

(1010, 150)

In [10]:
obj_list = list(data.select_dtypes(['object']).columns)
obj_list

['Smoking',
 'Alcohol',
 'Punctuality',
 'Lying',
 'Internet usage',
 'Gender',
 'Left - right handed',
 'Education',
 'Only child',
 'Village - town',
 'House - block of flats']

In [11]:
for i in obj_list:
    print(data[i].unique())

['never smoked' 'tried smoking' 'former smoker' 'current smoker']
['drink a lot' 'social drinker' 'never']
['i am always on time' 'i am often early' 'i am often running late']
['never' 'sometimes' 'only to avoid hurting someone'
 'everytime it suits me']
['few hours a day' 'most of the day' 'less than an hour a day'
 'no time at all']
['female' 'male']
['right handed' 'left handed']
['college/bachelor degree' 'secondary school' 'primary school'
 'masters degree' 'doctorate degree' 'currently a primary school pupil']
['no' 'yes']
['village' 'city']
['block of flats' 'house/bungalow']


I have derived that from 150 columns, my target column is 'Happiness in life'.

In [12]:
from sklearn.preprocessing import OrdinalEncoder

ord_enc = OrdinalEncoder()

for i in obj_list:
    data[i] = ord_enc.fit_transform(data[[i]])

In [13]:
data.head()

Music  Slow songs or fast songs  Dance  Folk  Country  Classical music  \
0    5.0                       3.0    2.0   1.0      2.0              2.0   
1    4.0                       4.0    2.0   1.0      1.0              1.0   
2    5.0                       5.0    2.0   2.0      3.0              4.0   
3    5.0                       3.0    2.0   1.0      1.0              1.0   
4    5.0                       3.0    4.0   3.0      2.0              4.0   

   Musical  Pop  Rock  Metal or Hardrock  ...   Age  Height  Weight  \
0      1.0  5.0   5.0                1.0  ...  20.0   163.0    48.0   
1      2.0  3.0   5.0                4.0  ...  19.0   163.0    58.0   
2      5.0  3.0   5.0                3.0  ...  20.0   176.0    67.0   
3      1.0  2.0   2.0                1.0  ...  22.0   172.0    59.0   
4      3.0  5.0   3.0                1.0  ...  20.0   170.0    59.0   

   Number of siblings  Gender  Left - right handed  Education  Only child  \
0                 1.0     0.0                  1.0        0.0         0.0   
1                 2.0     0.0                  1.0        0.0         0.0   
2                 2.0     0.0                  1.0        5.0         0.0   
3                 1.0     0.0                  1.0        0.0         1.0   
4                 1.0     0.0                  1.0        5.0         0.0   

   Village - town  House - block of flats  
0             1.0                     0.0  
1             0.0                     0.0  
2             0.0                     0.0  
3             0.0                     1.0  
4             1.0                     1.0  

[5 rows x 150 columns]

In [14]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1010 entries, 0 to 1009
Columns: 150 entries, Music to House - block of flats
dtypes: float64(145), int64(5)
memory usage: 1.2 MB


In [15]:
for k, v in data.items():
    q1 = v.quantile(0.25)
    q3 = v.quantile(0.75)
    iqr = q3 - q1
    v_col = v[(v <= q1 - 1.5 * iqr) | (v >= q3 + 1.5 * iqr)]
    perc = np.shape(v_col)[0] * 100.0 / np.shape(data)[0]
    if perc > 10:
        print("Column %s outliers = %.2f%%" % (k, perc))
#         if perc > 50:
#             data.drop(k)
#         else:
#             data=data[((data[k] < (q1-1.5*iqr)) | (data[k] > (q3+1.5*iqr)))]

Column Music outliers = 100.00%
Column Lying outliers = 13.66%
Column Left - right handed outliers = 100.00%
Column Education outliers = 20.99%


In [16]:
data = data.drop(['Left - right handed'], axis=1)
data = data.drop(['Music'], axis=1)

In [17]:
data.describe()

Slow songs or fast songs        Dance         Folk      Country  \
count               1010.000000  1010.000000  1010.000000  1010.000000   
mean                   3.327723     3.112871     2.287129     2.122772   
std                    0.833232     1.168268     1.136271     1.073501   
min                    1.000000     1.000000     1.000000     1.000000   
25%                    3.000000     2.000000     1.000000     1.000000   
50%                    3.000000     3.000000     2.000000     2.000000   
75%                    4.000000     4.000000     3.000000     3.000000   
max                    5.000000     5.000000     5.000000     5.000000   

       Classical music      Musical          Pop         Rock  \
count      1010.000000  1010.000000  1010.000000  1010.000000   
mean          2.956436     2.762376     3.473267     3.769307   
std           1.248223     1.259639     1.160029     1.185161   
min           1.000000     1.000000     1.000000     1.000000   
25%           2.000000     2.000000     3.000000     3.000000   
50%           3.000000     3.000000     4.000000     4.000000   
75%           4.000000     4.000000     4.000000     5.000000   
max           5.000000     5.000000     5.000000     5.000000   

       Metal or Hardrock         Punk  ...  Spending on healthy eating  \
count        1010.000000  1010.000000  ...                 1010.000000   
mean            2.357426     2.444554  ...                    3.558416   
std             1.372955     1.302356  ...                    1.092842   
min             1.000000     1.000000  ...                    1.000000   
25%             1.000000     1.000000  ...                    3.000000   
50%             2.000000     2.000000  ...                    4.000000   
75%             3.000000     3.000000  ...                    4.000000   
max             5.000000     5.000000  ...                    5.000000   

               Age       Height       Weight  Number of siblings       Gender  \
count  1010.000000  1010.000000  1010.000000          1010.00000  1010.000000   
mean     20.423762   173.444554    66.278218             1.29604     0.406931   
std       2.821521     9.936737    13.730769             1.01059     0.491505   
min      15.000000    62.000000    41.000000             0.00000     0.000000   
25%      19.000000   167.000000    56.000000             1.00000     0.000000   
50%      20.000000   172.000000    64.000000             1.00000     0.000000   
75%      22.000000   180.000000    75.000000             2.00000     1.000000   
max      30.000000   203.000000   165.000000            10.00000     1.000000   

         Education   Only child  Village - town  House - block of flats  
count  1010.000000  1010.000000     1010.000000             1010.000000  
mean      3.656436     0.251485        0.296040                0.406931  
std       2.012297     0.434082        0.456735                0.491505  
min       0.000000     0.000000        0.000000                0.000000  
25%       3.000000     0.000000        0.000000                0.000000  
50%       5.000000     0.000000        0.000000                0.000000  
75%       5.000000     1.000000        1.000000                1.000000  
max       5.000000     1.000000        1.000000                1.000000  

[8 rows x 148 columns]

In [18]:
data.shape

(1010, 148)

In [19]:
corr_matrix = data.corr().abs()
corr_series = corr_matrix['Happiness in life']
for i in corr_series.index:
    if corr_series[i] < 0.10:
        corr_series.drop(labels = [i], inplace = True)

In [20]:
x = list(corr_series.index)
x

['Dance',
 'Hiphop, Rap',
 'Reggae, Ska',
 'Mathematics',
 'Physics',
 'Economy Management',
 'Cars',
 'Active sport',
 'Fun with friends',
 'Adrenaline sports',
 'Fear of public speaking',
 'Healthy eating',
 'Prioritising workload',
 'Reliability',
 'Fake',
 'Criminal damage',
 'Self-criticism',
 'Eating to survive',
 'Loneliness',
 'Changing the past',
 'Dreams',
 'Number of friends',
 'New environment',
 'Mood swings',
 'Socializing',
 'Achievements',
 'Assertiveness',
 'Getting angry',
 'Knowing the right people',
 'Public speaking',
 'Life struggles',
 'Happiness in life',
 'Energy levels',
 'Personality',
 'Getting up',
 'Interests or hobbies',
 "Parents' advice",
 'Internet usage',
 'Entertainment spending',
 'Spending on healthy eating']

In [21]:
from sklearn.model_selection import train_test_split

x = data.drop(columns=x)
y = data['Happiness in life']

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.33, random_state=42)

In [22]:
print(x_train.shape) 
print(x_test.shape)
print(y_train.shape) 
print(y_test.shape)

(676, 108)
(334, 108)
(676,)
(334,)


In [23]:
from sklearn.linear_model import LogisticRegression
logreg = LogisticRegression()
logreg.fit(x_train, y_train)
logreg.classes_
y_preds_logreg = logreg.predict(x_test)
y_preds_logreg

c:\users\godfather\appdata\local\programs\python\python38\lib\site-packages\sklearn\linear_model\_logistic.py:938: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


array([4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 3., 4., 3., 4., 4., 3., 5.,
       4., 4., 3., 5., 4., 4., 4., 4., 4., 4., 4., 4., 3., 4., 4., 4., 3.,
       3., 5., 4., 4., 4., 4., 4., 3., 5., 5., 4., 3., 4., 4., 4., 3., 4.,
       4., 3., 3., 4., 4., 3., 3., 4., 3., 4., 4., 4., 4., 4., 4., 4., 4.,
       4., 3., 5., 4., 3., 4., 3., 4., 4., 3., 3., 4., 4., 4., 4., 3., 4.,
       4., 3., 3., 5., 4., 3., 3., 4., 4., 4., 3., 3., 3., 4., 3., 4., 5.,
       3., 4., 4., 4., 3., 4., 3., 3., 3., 4., 5., 3., 4., 4., 3., 3., 3.,
       3., 5., 3., 5., 4., 4., 4., 5., 2., 4., 5., 4., 3., 4., 4., 4., 4.,
       4., 4., 4., 3., 4., 4., 4., 4., 3., 5., 4., 4., 4., 4., 4., 4., 3.,
       4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 3., 3., 3., 4., 4., 4.,
       3., 4., 3., 3., 4., 3., 4., 4., 4., 4., 4., 4., 3., 3., 4., 3., 4.,
       4., 4., 4., 4., 3., 3., 4., 4., 3., 4., 5., 4., 4., 3., 4., 3., 5.,
       4., 3., 4., 4., 5., 4., 4., 4., 3., 3., 5., 3., 4., 4., 4., 4., 4.,
       3., 4., 3., 5., 4.

In [26]:
from sklearn import metrics

def mean_absolute_percentage_error(y_true, y_pred): 
    y_true, y_pred = np.array(y_true), np.array(y_pred)
    return np.mean(np.abs((y_true - y_pred) / y_true)) * 100

print('mean_absolute_error:')
print(metrics.mean_absolute_error(y_test,y_preds_logreg))
print('mean_squared_error:')
print(metrics.mean_squared_error(y_test,y_preds_logreg))
print('r2_score:')
print(metrics.r2_score(y_test,y_preds_logreg))
print('mean_absolute_percentage_error:')
print(mean_absolute_percentage_error(y_test,y_preds_logreg))
print('accuarcy score:')
print(metrics.accuracy_score(y_test,y_preds_logreg))

mean_absolute_error:
0.6497005988023952
mean_squared_error:
0.8892215568862275
r2_score:
-0.4166083541592287
mean_absolute_percentage_error:
22.080838323353294
accuarcy score:
0.45209580838323354
